raft/src/main/java/org/apache/kafka/raft/metadata/AbstractApiMessageSerde.java

```diff
@@ -0,0 +1,93 @@
+ *
+ * This can be used as the underlying serialization mechanism for records defined with {@link ApiMessage}s.
+ * <p></p>
+ * Serialization format for the given {@code ApiMessageAndVersion} is below:
+ * <p></p>
+ * <pre>
+ *     [data_frame_version header message]
+ *     header => [api_key version]
+ *
+ *     data_frame_version   : This is the header version, current value is 0. Header includes both api_key and version.
+ *     api_key              : apiKey of {@code ApiMessageAndVersion} object.
+ *     version              : version of {@code ApiMessageAndVersion} object.
+ *     message              : serialized message of {@code ApiMessageAndVersion} object.
+ * </pre>
+ */
+public abstract class AbstractApiMessageSerde implements RecordSerde<ApiMessageAndVersion> {
+    private static final short DEFAULT_FRAME_VERSION = 0;
+    private static final int DEFAULT_FRAME_VERSION_SIZE = ByteUtils.sizeOfUnsignedVarint(DEFAULT_FRAME_VERSION);
+
+    @Override
+    public int recordSize(ApiMessageAndVersion data,
+                          ObjectSerializationCache serializationCache) {
+        int size = DEFAULT_FRAME_VERSION_SIZE;
+        size += ByteUtils.sizeOfUnsignedVarint(data.message().apiKey());
+        size += ByteUtils.sizeOfUnsignedVarint(data.version());
+        size += data.message().size(serializationCache, data.version());
+        return size;
+    }
+
+    @Override
+    public void write(ApiMessageAndVersion data,
+                      ObjectSerializationCache serializationCache,
+                      Writable out) {
+        out.writeUnsignedVarint(DEFAULT_FRAME_VERSION);
+        out.writeUnsignedVarint(data.message().apiKey());
+        out.writeUnsignedVarint(data.version());
+        data.message().write(out, serializationCache, data.version());
+    }
+
+    @Override
+    public ApiMessageAndVersion read(Readable input,
+                                     int size) {
+        short frameVersion = (short) input.readUnsignedVarint();
+        if (frameVersion != DEFAULT_FRAME_VERSION) {
+            throw new SerializationException("Could not deserialize metadata record due to unknown frame version "
+                                                     + frameVersion + "(only frame version " + DEFAULT_FRAME_VERSION + " is supported)");
+        }
+
+        short apiKey = (short) input.readUnsignedVarint();
+        short version = (short) input.readUnsignedVarint();
+        ApiMessage record = apiMessageFor(apiKey);
+        record.read(input, version);
+        return new ApiMessageAndVersion(record, version);
+    }
+
+    /**
+     * Return {@code ApiMessage} instance for the given {@code apiKey}. This is used while deserializing the bytes
+     * payload into the respective {@code ApiMessage} in {@link #read(Readable, int)} method.
+     *
+     * @param apiKey apiKey for which a {@code ApiMessage} to be created.
+     */
+    public abstract ApiMessage apiMessageFor(short apiKey);
```

Comment:
Hello @satishd , why should we make this abstract? I find that the two implementations of this method are all `MetadataRecordType.fromId(apiKey).newMetadataRecord()`.

Reply:
@dengziming Below are the two implementations of `AbstractApiMessageSerde` and they do not share same records types.
- [BytesApiMessageSerde](https://github.com/apache/kafka/blob/trunk/server-common/src/main/java/org/apache/kafka/server/common/serialization/BytesApiMessageSerde.java#L41rl)
```BytesApiMessageSerde.this.apiMessageFor(apiKey);```

- [MetadataRecordSerde](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/MetadataRecordSerde.java#L27)
```MetadataRecordType.fromId(apiKey).newMetadataRecord();```